### PR TITLE
Datagenerator library

### PR DESCRIPTION
**Pull request by dtarakanov1**

### CODE REVIEW

File: BaikalDataGenerator/Source/render.h [hunk 1]
```diff
@@ -51,120 +60,54 @@ class Render
...
-    void GenerateSample(const CameraInfo& cam_state,
-                        const std::vector<size_t>& spp,
-                        const std::filesystem::path& output_dir,
-                        bool gamma_correction_enabled,
-                        size_t start_cam_id);
-
-    void SaveMetadata(const std::filesystem::path& output_dir) const;
-
-    std::filesystem::path m_scene_file;
+    SceneObject* m_scene = nullptr;
```
Comment (AvKhokhlov) [inline BaikalDataGenerator/Source/render.h]: Why not unique_ptr. Yes, I know about the problem with destructor, but we alreadyhave workaround.

Comment (AvKhokhlov) [reply BaikalDataGenerator/Source/render.h]: or shared_ptr

File: BaikalStandalone/Application/cl_render.cpp [hunk 2]
```diff
@@ -71,7 +71,10 @@ namespace Baikal
 
         for (std::size_t i = 0; i < m_cfgs.size(); ++i)
         {
-            std::cout << i << ": " << m_cfgs[i].context.GetDevice(0).GetName() << "\n";
+            std::cout << i << ". name: " << m_cfgs[i].context.GetDevice(0).GetName()
```
Comment (AvKhokhlov) [inline BaikalStandalone/Application/cl_render.cpp]: Make some variable. For example  some_variable  = m_cfgs[i].context.GetDevice(0).

File: CMakeLists.txt [hunk 3]
```diff
@@ -7,11 +7,11 @@ option(BAIKAL_ENABLE_DENOISER "Use denoising on output" OFF)
 option(BAIKAL_ENABLE_RPR "Enable RadeonProRender API lib" OFF)
 option(BAIKAL_ENABLE_TESTS "Enable tests" ON)
 option(BAIKAL_ENABLE_STANDALONE "Enable standalone application build" ON)
-option(BAIKAL_ENABLE_DATAGENERATOR "Enable data generator application build" OFF)
+option(BAIKAL_ENABLE_DATAGENERATOR "Enable data generator application build" ON)
 option(BAIKAL_ENABLE_IO "Enable IO library build" ON)
 option(BAIKAL_ENABLE_FBX "Enable FBX import in BaikalIO. Requires BaikalIO to be turned ON" OFF)
 option(BAIKAL_ENABLE_MATERIAL_CONVERTER "Enable materials.xml converter from old to uberv2 version" OFF)
-option(BAIKAL_EMBED_KERNELS "Embed CL kernels into binary module" OFF)
+option(BAIKAL_EMBED_KERNELS "Embed CL kernels into binary module" ON)
```
Comment (AvKhokhlov) [inline CMakeLists.txt]: I'm not sure that it should be enabled by default.

File: Baikal/Controllers/clw_scene_controller.cpp [hunk 4]
```diff
@@ -227,7 +227,7 @@ namespace Baikal
             auto transform = mesh->GetTransform();
             shape->SetTransform(transform, inverse(transform));
             shape->SetId(id++);
-            shape->SetMask(iter->GetVisibilityMask());
+            // shape->SetMask(iter->GetVisibilityMask());
```
Comment (AvKhokhlov) [inline Baikal/Controllers/clw_scene_controller.cpp]: Are you sure we don't need this?

File: BaikalDataGenerator/Source/data_generator.h [hunk 5]
```diff
@@ -22,35 +22,54 @@ THE SOFTWARE.
...
-using CameraIterator = std::vector<CameraInfo>::const_iterator;
-using LightsIterator = std::vector<LightInfo>::const_iterator;
-using SppIterator = std::vector<int>::const_iterator;
-
-class ConfigLoader
+struct DataGeneratorParams
 {
-public:
-    explicit ConfigLoader(const DGenConfig& config);
+    const char* output_dir;
```
Comment (AvKhokhlov) [inline BaikalDataGenerator/Source/data_generator.h]: char const*

File: BaikalDataGenerator/Source/main.cpp [hunk 6]
```diff
@@ -21,55 +21,30 @@ THE SOFTWARE.
...
-    }
-
-    render.GenerateDataset(camera_states_subset,
-                           config_loader.Lights(),
-                           config_loader.LightsDir(),
-                           config_loader.Spp(),
-                           config.output_dir,
-                           config.gamma_correction,
-                           config.offset_idx);
+    ObjectLoader config_loader(config);
```
Comment (AvKhokhlov) [inline BaikalDataGenerator/Source/main.cpp]: Maybe ObjectLoader object_loader would be better?

File: BaikalDataGenerator/Source/object_loader.cpp [hunk 7]
```diff
@@ -0,0 +1,392 @@
...
+{
+    if (!std::filesystem::exists(m_app_config.scene_file))
+    {
+        THROW_EX("There is no any scene file to load");
+    }
+
+    // workaround to avoid issues with tiny_object_loader
+    auto scene_dir = m_app_config.scene_file.parent_path().string();
+
+    if (scene_dir.back() != '/' || scene_dir.back() != '\\')
```
[the remark this answers is not in the document]
Comment (dtarakanov1) [reply BaikalDataGenerator/Source/object_loader.cpp]: `if` is not needed here, but the trailing delimiter is required.

File: BaikalDataGenerator/Source/render.h (hunk 1 above)
Comment (dtarakanov1) [reply BaikalDataGenerator/Source/render.h]: The render does not own the scene, a client does. A scene is passed through the shared object boundary as `rpr_scene` (= `void*`), it's not a good idea to wrap an extern pointer into a smart pointer here.

File: BaikalDataGenerator/Source/object_loader.cpp [hunk 8]
```diff
@@ -0,0 +1,392 @@
...
+    size_t end = 0;
+
+    if (subrange_idx < remain)
+    {
+        begin = subrange_idx * (length + 1);
+        end = begin + length + 1;
+    }
+    else
+    {
+        begin = remain * (length + 1) + (subrange_idx - remain) * length;
```
Comment (AvKhokhlov) [inline BaikalDataGenerator/Source/object_loader.cpp]: same as begin = subrange_idx * length + remain

File: BaikalDataGenerator/Source/render.cpp [hunk 9]
```diff
@@ -138,179 +121,90 @@ Render::Render(const std::filesystem::path& scene_file,
...
-                                                     cam_state.pos,
-                                                     cam_state.up);
-
-        // default sensor width
-        float sensor_width = 0.036f;
-        float sensor_height = static_cast<float>(m_height) / static_cast<float>(m_width) * sensor_width;
-
-        m_camera->SetSensorSize(float2(0.036f, sensor_height));
-        m_camera->SetDepthRange(float2(0.0f, 100000.f));
+    m_scene->SetCamera(camera);
```
[the remark this answers is not in the document]
Comment (dtarakanov1) [reply BaikalDataGenerator/Source/render.cpp]: The check is done in the `data_generator.cpp` right before the call.

File: Baikal/Controllers/clw_scene_controller.cpp (hunk 4 above)
Comment (dtarakanov1) [reply Baikal/Controllers/clw_scene_controller.cpp]: CI fails to build with this line uncommented.

File: BaikalDataGenerator/Source/data_generator.cpp [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+    }
+
+    auto* scene = SceneObject::Cast<SceneObject>(params->scene);
+    if (scene == nullptr)
+    {
+        return kDataGeneratorBadScene;
+    }
+
+    Render render(scene,
+                  params->width,
```
Comment (AvKhokhlov) [inline BaikalDataGenerator/Source/data_generator.cpp]: We need some checks for zeros

File: BaikalDataGenerator/Source/data_generator.cpp [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+        if (light == nullptr)
+        {
+            return kDataGeneratorBadLights;
+        }
+        render.AttachLight(light);
+    }
+
+    unsigned camera_end_idx = params->cameras_start_idx + params->cameras_num - 1;
+
+    render.SaveMetadata(output_dir,
```
Comment (yozhijk) [inline BaikalDataGenerator/Source/data_generator.cpp]: Would be good to add some comments on the most important implementation parts

File: BaikalDataGenerator/Source/object_loader.cpp [hunk 12]
```diff
@@ -0,0 +1,390 @@
...
+        THROW_EX("'split_num' should be positive");
+    }
+
+    if (config.split_idx >= config.split_num)
+    {
+        THROW_EX("'split_idx' must be less than split_num");
+    }
+}
+
+std::pair<std::size_t, std::size_t> GetSplitByIdx(std::size_t total_num,
```
Comment (yozhijk) [inline BaikalDataGenerator/Source/object_loader.cpp]: I would return a struct with human readable field names, pair is a bit confusing on the client side

Comment (yozhijk) [reply BaikalDataGenerator/Source/object_loader.cpp]: or at least using range = std::pair<size_t, size_t>, and range.begin, range.end would make much more sense

File: BaikalDataGenerator/Source/data_generator.cpp [hunk 13]
```diff
@@ -0,0 +1,163 @@
...
+
+#include "devices.h"
+#include "render.h"
+#include "utils.h"
+
+#include "Rpr/WrapObject/LightObject.h"
+#include "Rpr/WrapObject/SceneObject.h"
+
+
+DataGeneratorResult GenerateDataset(DataGeneratorParams const* params)
```
Comment (achernigin1987) [inline BaikalDataGenerator/Source/data_generator.cpp]: may be GenerateDataset -> bdgGenerateDataset ?

File: BaikalDataGenerator/Source/render.h [hunk 14]
```diff
@@ -51,120 +60,54 @@ class Render
...
+    void SaveMetadata(const std::filesystem::path& output_dir,
+                      const std::string& scene_name,
+                      unsigned cameras_start_idx,
+                      unsigned cameras_end_idx,
+                      int cameras_index_offset,
+                      bool gamma_correction_enabled) const;
+
+    void GenerateSample(CameraObject* camera,
+                        int camera_idx,
+                        const std::vector<size_t>& spp,
```
Comment (achernigin1987) [inline BaikalDataGenerator/Source/render.h]: size_t -> unsigned

File: BaikalDataGenerator/Source/render.h [hunk 15]
```diff
@@ -22,27 +22,36 @@ THE SOFTWARE.
```
Comment (achernigin1987) [inline BaikalDataGenerator/Source/render.h]: lets rename file and class to data_generator_impl and DataGeneratorImpl correspondingly

File: BaikalDataGenerator/Source/data_generator.cpp [hunk 16]
```diff
@@ -0,0 +1,163 @@
...
+    {
+        return kDataGeneratorBadLights;
+    }
+
+    if (params->spp == nullptr || params->spp_num == 0)
+    {
+        return kDataGeneratorBadSpp;
+    }
+    // Sort SPP list and remove duplicates
+    std::vector<size_t> sorted_spp(params->spp, params->spp + params->spp_num);
```
Comment (achernigin1987) [inline BaikalDataGenerator/Source/data_generator.cpp]: size_t -> unsigned

File: BaikalDataGenerator/Source/data_generator.h [hunk 17]
```diff
@@ -22,35 +22,59 @@ THE SOFTWARE.
...
-    explicit ConfigLoader(const DGenConfig& config);
+    char const* output_dir;
+
+    rpr_scene scene;
+    char const* scene_name;
+
+    rpr_camera* cameras;
+    unsigned cameras_start_idx;
+    unsigned cameras_num;
+    int cameras_offset_idx;
```
Comment (achernigin1987) [inline BaikalDataGenerator/Source/data_generator.h]: Lets remove cameras_offset_idx from here and generate images for all passed cameras, starting output index from cameras_start_idx. Also lets split cameras outside datagenerator implementation - in the our application part

File: BaikalDataGenerator/Source/render.cpp [hunk 18]
```diff
@@ -138,179 +121,90 @@ Render::Render(const std::filesystem::path& scene_file,
...
-    // log outputs layout
-    XMLElement* size_attribute = doc.NewElement("layout");
-    size_attribute->SetAttribute("width", m_width);
-    size_attribute->SetAttribute("height", m_height);
-    root->InsertEndChild(size_attribute);
+    auto* cameras = doc.NewElement("cameras");
+    cameras->SetAttribute("start_idx", static_cast<int>(cameras_start_idx));
+    cameras->SetAttribute("end_idx", static_cast<int>(cameras_end_idx));
+    cameras->SetAttribute("idx_offset", cameras_index_offset);
+    root->InsertEndChild(cameras);
```
Comment (achernigin1987) [inline BaikalDataGenerator/Source/render.cpp]: After removing parameter cameras_start_idx we don't need node "cameras" at all. But we have to put cameras info into dedicated xml file (i.e. app_metadata.xml) to be able handle it on the server side.

File: BaikalDataGenerator/Source/data_generator.cpp (hunk 16 above)
Comment (AvKhokhlov) [reply BaikalDataGenerator/Source/data_generator.cpp]: Done

File: BaikalDataGenerator/Source/render.h (hunk 14 above)
Comment (AvKhokhlov) [reply BaikalDataGenerator/Source/render.h]: Done

**Pull request merged**